### PR TITLE
Revamp Table Catalog, Add Environment Vars for Table Configuration

### PR DESCRIPTION
**Pull request by bitner**
## Updates Table Introspection Query
- Organize structure for performance
- Add calls to has_table_privilege and has_column_privilege to ensure permissions check
- Use postgis_typmod_type and postgis_typmod_srid rather than geometry_columns/geography_columns tables
- Allow for null primary key to be returned if a Primary Key or Unique Key are not found
- Return datetime columns from database query

## Add Environment Variables for Table Settings
- TIFEATURES_FALLBACK_KEY_NAMES -- takes a list of column names that will be considered a primary key without any unique constraints (constraints should always be added to Tables, but Views cannot have unique constraints, so this gives a method to always return a standard column name in a view if that should be used as a primary key)
- TIFEATURES__TABLE_CONFIG allows to add configuration for a specific table for what should be used as the primary key, default geometry and datetime column and stubs in support for setting spatial and temporal extents.
  - Example to define a primary key column for a table named test (note that public.test needs to be specified as public_test) `TIFEATURES__TABLE_CONFIG__public_test__pk=gid`
  - Example to set the default geometry column on a table that may have multiple geometry columns `TIFEATURES__TABLE_CONFIG__public_test__geomcol=geometry`

## Update logic for dealing with primary key
If a primary key cannot be found either by table constraints, fallback_key_names, or table_config, a table will be made available to the /collections/public.table/items endpoint, but individual items will not be able to be accessed via /collections/public.table/items/itemid

### CODE REVIEW

File: tifeatures/dbmodel.py [hunk 1]
```diff
@@ -48,33 +52,69 @@ def json_type(self) -> str:
...
-    name: str
-    bounds: List[float]
-    srid: int
+    bounds: List[float] = [-180, -90, 180, 90]
+    srid: int = 4326
     geometry_type: str
 
 
+class DatetimeColumn(Column):
+    """Model for PostGIS geometry/geography column."""
```
Comment (vincentsarago) [inline tifeatures/dbmodel.py:64]:
```suggestion
    """Model for PostGIS datetime column."""
```

File: tifeatures/dbmodel.py [hunk 2]
```diff
@@ -48,33 +52,69 @@ def json_type(self) -> str:
...
+    def validate_list_columns(cls, v):
+        """Validate List Columns."""
+        if v is None:
+            return []
+        return v
 
     def datetime_column(self, name: Optional[str] = None) -> Optional[Column]:
         """Return the Column for either the passed in tstz column or the first tstz column."""
+        if self.datetime_columns is None or len(self.datetime_columns) == 0:
+            return None
```
[the remark this answers is not in the document]
Comment (bitner) [reply tifeatures/dbmodel.py:114]: Is empty list falsey?? If so, yeah

File: tifeatures/dbmodel.py [hunk 3]
```diff
@@ -83,9 +123,16 @@ def datetime_column(self, name: Optional[str] = None) -> Optional[Column]:
...
+        if self.geometry_columns is None or len(self.geometry_columns) == 0:
+            return None
         if name and name.lower() == "none":
             return None
-
+        if name is None:
+            name = (
+                getattr(self.table_config, "geomcol", None)
+                if self.table_config
+                else None
```
Comment (vincentsarago) [inline tifeatures/dbmodel.py:134]: why not do like in line 117?

File: tifeatures/dbmodel.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vincentsarago) [reply tifeatures/dbmodel.py:114]: yes

File: tifeatures/filter/filters.py [hunk 4]
```diff
@@ -17,13 +17,13 @@
 
 def bbox_to_wkt(bbox: List[float], srid: int = 4326) -> str:
     """Return WKT representation of a BBOX."""
-    poly = Polygon.from_bounds(*bbox)
+    poly = Polygon.from_bounds(*bbox)  # type:ignore
```
Comment (vincentsarago) [inline tifeatures/filter/filters.py:20]: 😅

File: tifeatures/layer.py [hunk 5]
```diff
@@ -9,7 +9,7 @@
 from buildpg import funcs as pg_funcs
 from buildpg import logic, render
 from ciso8601 import parse_rfc3339
-from geojson_pydantic import Feature, FeatureCollection
+from geojson_pydantic import Feature, FeatureCollection  # type:ignore
```
Comment (vincentsarago) [inline tifeatures/layer.py:12]: what is the issue here?

File: tifeatures/layer.py [hunk 6]
```diff
@@ -427,13 +430,17 @@ async def query(
...
+            else pg_funcs.cast(None, "text"),
             geometry_q=self._geom(
                 geometry_column=self.geometry_column(geom),
                 bbox_only=bbox_only,
                 simplify=simplify,
             ),
-            geom_columns=[g.name for g in self.geometry_columns],
+            geom_columns=[g.name for g in self.geometry_columns]
+            if self.geometry_columns
+            else [],
```
Comment (vincentsarago) [inline tifeatures/layer.py:443]: can we extract the if/else blocks, IMO it makes the code hard to follow

File: tifeatures/layer.py [hunk 7]
```diff
@@ -492,10 +499,13 @@ async def feature(
     @property
     def queryables(self) -> Dict:
         """Return the queryables."""
-        geoms = {
-            col.name: {"$ref": geojson_schema.get(col.geometry_type.upper(), "")}
-            for col in self.geometry_columns
-        }
+        if self.geometry_columns and len(self.geometry_columns) > 0:
```
Comment (vincentsarago) [inline tifeatures/layer.py:502]: same 
```python
if not self.geometry_columns:
```

**Review by vincentsarago — COMMENTED** [review]
Looks good to me, maybe we need more tests then!

File: tifeatures/layer.py (hunk 5 above)
Comment (vincentsarago) [reply tifeatures/layer.py:12]: 🤔 https://github.com/developmentseed/geojson-pydantic/blob/master/geojson_pydantic/__init__.py#L17

can you check locally if you do `from geojson_pydantic.features import Feature, FeatureCollection`

File: tifeatures/settings.py [hunk 8]
```diff
@@ -21,6 +21,7 @@ class _APISettings(pydantic.BaseSettings):
     """API settings"""
 
     name: str = "TiFeatures"
+    DEBUG: bool = False
```
Comment (vincentsarago) [inline tifeatures/settings.py:24]: 👀 ?

Comment (bitner) [reply tifeatures/settings.py:24]: It makes it much easier to debug what's getting into the catalog

File: tests/routes/test_items.py [hunk 9]
```diff
@@ -747,3 +747,41 @@ def test_items_sortby(app):
...
+    )
+    body = response.json()
+    assert body["features"][0]["id"] == "1"
+
+    monkeypatch.setenv("TIFEATURES_TABLE_CONFIG__public_my_data__pk", "id")
+    response = app.get(
+        "/collections/public.my_data/items?datetime=2004-10-19T10:23:54Z"
+    )
+    body = response.json()
+    assert body["features"][0]["id"] == "0"
```
Comment (vincentsarago) [inline tests/routes/test_items.py:787]: ❤️

File: tifeatures/dbmodel.py [hunk 10]
```diff
@@ -70,68 +68,39 @@ class DatetimeColumn(Column):
...
+            return self.datetime_column
         for col in self.datetime_columns:
-            if name is None or col.name == name:
+            if col.name == name:
                 return col
-
         return None
 
-    def geometry_column(self, name: Optional[str] = None) -> Optional[GeometryColumn]:
+    def get_geometry_column(
```
Comment (vincentsarago) [inline tifeatures/dbmodel.py:93]: ✨
